functions/src/update-stats.ts

```diff
@@ -24,47 +20,50 @@ export const updateStats = functions.firestore
       return;
     }
     const vote = voteSnapshot.data()! as Vote;
-    let stats: StatsData;
+    const voteId: string = voteSnapshot.id
     await firebase.firestore().runTransaction(async transaction => {
-      await transaction
-        .get(
+      const doc = await transaction.get(
+        firebase
+          .firestore()
+          .collection("stats")
+          .doc(vote.campaign)
+      );
+      if (doc.get(voteId)) {
+        console.log("this vote has already been counted");
+        throw new functions.https.HttpsError(
+          "already-exists",
+          `This vote has already been counted`
+        );
+      }
+
+      let data = doc.data();
+      if (!data) {
+        data = {};
+        data[vote.value] = 1;
+      } else {
+        if (!data[vote.value]) {
+          data[vote.value] = 1;
+        } else {
+          data[vote.value] += 1;
+        }
+      }
```

Comment:
Same remark as with `stats` before. Try to build up your data structure in one go to avoid state, ie `const data = ...` and never modify `data` again. Also, try to find a more descriptive name for `data`. :)